src/python/vdo-setup-dmtest/source/Utils.py

```diff
@@ -0,0 +1,125 @@
+        distinguished_name = req_distinguished_name
+        prompt = no
+        string_mask = utf8only
+        x509_extensions = myexts
+        [ req_distinguished_name ]
+        CN = Build time autogenerated kernel key for test modules
+        [ myexts ]
+        basicConstraints=critical,CA:FALSE
+        keyUsage=digitalSignature
+        subjectKeyIdentifier=hash
+        authorityKeyIdentifier=keyid
+        """))
+
+    result = runCommand(['sudo cp', genkeyFile, certsDir], logger)
+    assert result.returncode == 0, " ".join(['Copying', filename, 'to', certsDir, 'failed'])
+  except Exception as error:
+    logger.error("Error - {0}".format(error))
+    return False
+
+  return True
+
+def isExecutable(filePath):
+  """
+  Check if the file at the specified path is executable
+  """
+  return os.path.isfile(filePath) and os.access(filePath, os.X_OK)
+
+def isURL(inputString):
+  """
+  Check if the input string is a URL
+  """
+  # The validators utility cannot evaluate git SSH URLs (nor can the Python standard urllib.parse).
+  # Similar to urllib.parse, it only verifies that the input has the proper format of a URL. It is
+  # an improvement over urllib.parse because it will catch if there are an invalid number of '/'
+  # between the URL scheme and netloc, along with whether the scheme is invalid.
+  if inputString.startswith('git@'):
+    # Return true for git SSH URLs. gitPython will evaluate the url downstream.
+    return True
+  else:
+    valid = validators.url(inputString)
+    if not valid:
+      return False
+  return True
+
+def runCommand(args, ignoreErrors=False):
+  """
+  Run a shell command
+  """
+  if isinstance(args, str):
+    command = args
+  else:
+    command = " ".join(args)
+  
+  logger.debug("Running Command: {0}".format(command))
+  result = subprocess.run(command, shell=True, text=True, capture_output=True)
+
+  if (not ignoreErrors) and (result.returncode != 0):
+    logger.error("Command '{cmd}' failed: stderr:\n{err}\nstdout:\n{out}".format(
+      cmd = command, err = result.stderr, out = result.stdout))
+  elif result.stdout != "":
+    logger.debug("runCommand stdout:\n{0}".format(result.stdout.rstrip()))
+  return result
+
+def runCommandIgnoringErrors(args):
+  """
+  Run a shell command ignoring any errors
+  """
+  return runCommand(args, True)
+
+def verifyCertificate(certPath):
+  """
+  Verify that a specified module signing certificate is valid and not expired
+  """
+  logger.info("Verifying the signing certificate validity")
+  daySeconds = str(86400)
+  valid = False
+
+  # Evaluate if the certificate will expire in the next day
+  try:
+    result = runCommandIgnoringErrors(['sudo openssl x509 -checkend', daySeconds, '-noout -in',
+                                       certPath])
+    if ('will not expire' in result.stdout):
+      valid = True
+    elif ('will expire' in result.stdout):
+      logger.info("Certificate will expire within 24 hours - recreation needed")
```

Comment:
Is there any significance to 24 hours?  How long is reasonable for dmtest to actually run?  Can it exceed 24 hours?

Perhaps `recreation needed` can be rephrase?  I like to play Dodgeball, but I don't think it'll help here ;)

Maybe `please re-generate certificate` ??

Reply:
There is no significance to 24 hours beyond giving it a date to compare to for determining certificate validity.

As for the "recreation needed", that can be rephrased, but we regenerate the certificate from within vdo-setup-dmtest in that case, so "please re-generate certificate" wouldn't be appropriate. Maybe "Certificate will expire within 24 hours - regenerating".

Reply:
Updated the message in the latest push to note that it's regenerating the certificate when it detects it will expire within a day.